### ISSUE TITLE
Would the current Loss Function work for non-square images?

### ISSUE THREAD

**Issue by datlife** [allanzelener/YAD2K#51]
Hi @allanzelener , 

Thank you for your great work. I wonder if  the current Loss Function works for non-square images?

In file `keras_yolo.py`, I see this:
```
    feats      = K.reshape(feats, [-1, conv_dims[0], conv_dims[1], num_anchors, num_classes + 5])
    conv_dims  = K.cast(K.reshape(conv_dims, [1, 1, 1, 1, 2]), K.dtype(feats))

    box_xy          = K.sigmoid(feats[..., :2]) 
    box_wh          = K.exp(feats[..., 2:4])
    box_confidence  = K.sigmoid(feats[..., 4:5])
    box_class_probs = K.softmax(feats[..., 5:])
```

I am trying to figure out if breaking `box_xy` --> `box_x` and `box_y` would allow me to use non-square image. Is there any other places in your code I need to be careful if I do that?

Thanks, :)

**Comment by allanzelener** [comment]
No, separating x and y is not the issue here. I don't remember exactly why
rectangular images were broken but I think how `conv_index` is computed is
the most likely culprit.

**Comment by datlife** [comment]
Thank you for you prompt response, @allanzelener ! I will report again if I could figure out how to tweak the code to work on non-square images.

**Comment by datlife** [comment]
Just out of curiosity, what do `con_dims` and `conv_index` mean (below)? In the paper, the author gave an example of how the predicted box is calculated as:
```
bx = sigmoid(tx) + cx 
by = sigmod(ty) + cy 
```
where as:
* `(cx, cy)` is the offset of a grid cell  from top left corner of the image
* `(tx, ty)`  is  prediction from the network.

In your Loss Implementation, is there any relationship between `conv_index` and `(cx, cy)`? 

Also, is there a reason why you decided to divide the whole thing by `conv_dims`?  Was it used to describe the location of a gird cell or were it for adjusting/scaling the box in certain way? 

```# Adjust preditions to each spatial grid point and anchor size.
# Note: YOLO iterates over height index before width index.
box_xy = (box_xy + conv_index) / conv_dims
....
```


I am trying to wrap my head around your code. Any help/comment would be greatly appreciated!!

Regards,

**Comment by allanzelener** [comment]
`conv_dims` are the dimensions of the final convolutional layer. If the input image size is (416, 416) then conv_dims are (13, 13). `conv_index` contains every possible index into the convolution layer, so in this example it's all pairs of numbers from 0 - 12. 

The model is trained so that each prediction for an index into the conv features is given as an offset from the top left corner of the corresponding cell in the original image. The conv_index is added to the predictions to make them relative to the top left corner of the whole image and dividing by conv_dims normalizes the prediction to [0, 1] so that it's represented as a percentage of the original image size.

`conv_index` may be incorrect for rectangular images because it uses some not so straightforward tensor manipulations to explicitly get all the indices. The original Darknet code just used for loops for these and is more readable but that's not easy to implement in TF/Keras either.

**Comment by datlife** [comment]
Hi @allanzelener ,

You are correct, the `conv_index` is out of order when the image size is not square. I found a way to create a properly `conv_index` with non-square images. Here is my proposed fix.

```
conv_dims = [4 , 5]  # example

# Create grid map
cx = tf.cast((K.arange(0, stop=conv_dims[1])), dtype=tf.float32)
cx = K.tile(cx, [conv_dims[0]])
cx = K.reshape(cx, [-1, conv_dims[0], conv_dims[1], 1])

cy = K.cast((K.arange(0, stop=conv_dims[1])), dtype=tf.float32)
cy = K.reshape(cy, [-1, 1])
cy = K.tile(cy, [1, conv_dims[0]])  
cy = K.reshape(cy, [-1])    
cy = K.reshape(cy, [-1, conv_dims[0], conv_dims[1], 1])

c_xy = tf.stack([cx, cy], -1)
c_xy = K.cast(c_xy, K.dtype(feats))

## This is the old one
# In YOLO the height index is the inner most iteration.
conv_height_index = K.arange(0, stop=conv_dims[0])
conv_height_index = K.tile(conv_height_index, [conv_dims[1]])

# TODO: Repeat_elements and tf.split doesn't support dynamic splits.
conv_width_index = K.arange(0, stop=conv_dims[1])
conv_width_index = K.tile(K.expand_dims(conv_width_index, 0), [conv_dims[0], 1])
conv_width_index = K.flatten(K.transpose(conv_width_index))
conv_index = K.transpose(K.stack([conv_height_index, conv_width_index]))
conv_index = K.reshape(conv_index, [1, conv_dims[0], conv_dims[1], 1, 2])
conv_index = K.cast(conv_index, K.dtype(feats))
```
Now, let's see what happens when I run this one with your current one.
```
with tf.Session() as sess:
    print(conv_index.eval())
    print(c_xy.eval())
```

**OLD ONE**  It produces incorrect `height_index` with non-square images
```
[[[[[ 0.  0.]]

   [[ 1.  0.]]

   [[ 2.  0.]]

   [[ 3.  0.]]

   [[ 0.  1.]]]

....

  [[[ 3.  3.]]

   [[ 0.  4.]]

   [[ 1.  4.]]

   [[ 2.  4.]]

   [[ 3.  4.]]]]]
```
**NEW ONE**
```
[[[[[ 0.  0.]]

   [[ 1.  0.]]

   [[ 2.  0.]]

   [[ 3.  0.]]

   [[ 4.  0.]]]

...
  [[[ 0.  3.]]

   [[ 1.  3.]]

   [[ 2.  3.]]

   [[ 3.  3.]]

   [[ 4.  3.]]]]]
```
Hope it would  help!

**Comment by datlife** [comment]
However, I am running into an a new issue. How can I port the pre-trained weights of squared input size, saying `408, 408`, into a new model with different input size (`1280, 960` in this case).

**Comment by fferroni** [comment]
Did you test the non-square modification?
I ported the weights by adding the --fully_convolutional flag (`./yad2k.py yolo.cfg yolo.weights model_data/yolo.h5 --fully_convolutional`) so that it does not require known input size (although you can specify this in `model_image_size` and `is_fixed_size` later in testing, but I get weird positions of bounding boxes even using your new conv_index function in `./test.py`

**Comment by tianyu-tristan** [comment]
Does this discussion have any conclusion? I'm also working with non-squared image and I'd like to use this great effort as YOLOv2 implementation, so far my overfitting on 5 samples gives me good loss but weird bounding boxes...

**Comment by tianyu-tristan** [comment]
@datlife btw there's a fix on the issue you stated above: https://github.com/allanzelener/YAD2K/issues/86

**Comment by tianyu-tristan** [comment]
I finally got my non-square image training working. Few things to change:

(1) conv_index (ref: https://github.com/allanzelener/YAD2K/blob/master/yad2k/models/keras_yolo.py#L112) is wrong as @datlife mentioned above. The solution @datlife provided above does not directly solve my problem, a slight change of that did:

    # add below conv_width_index = K.flatten(K.transpose(conv_width_index))
    # note that at this point conv_dims still follows [height, width]
    enum_width = tf.cast((K.arange(0, stop=conv_dims[1])), dtype=tf.float32)
    enum_width = K.tile(enum_width, [conv_dims[0]])
    enum_width = K.reshape(enum_width, [-1, conv_dims[0], conv_dims[1], 1])
    enum_height = K.cast((K.arange(0, stop=conv_dims[0])), dtype=tf.float32)
    enum_height = K.reshape(enum_height, [-1, 1])
    enum_height = K.tile(enum_height, [1, conv_dims[1]])
    enum_height = K.reshape(enum_height, [-1])
    enum_height = K.reshape(enum_height, [-1, conv_dims[0], conv_dims[1], 1])
    conv_index = tf.stack([enum_width, enum_height], -1)

    # conv_index = K.transpose(K.stack([conv_height_index, conv_width_index]))
    conv_index = K.reshape(conv_index, [1, conv_dims[0], conv_dims[1], 1, 2])
    conv_index = K.cast(conv_index, K.dtype(feats))

The basic idea is that you want `conv_index[0, 15, 8, 0, :] == [8, 15]` and `conv_index[0, 12, 6, 0, :] == [6,12]` so that later part `box_xy = (box_xy + conv_index) / conv_dims` will make sense. 

(2) conv_dims need to be fixed to `[width, height]`, solution mentioned in https://github.com/allanzelener/YAD2K/issues/86 to add one line of code.

Thanks all for discussing the issue here

**Comment by orangebacked** [comment]
> I finally got my non-square image training working. Few things to change:
> 
> (1) conv_index (ref: https://github.com/allanzelener/YAD2K/blob/master/yad2k/models/keras_yolo.py#L112) is wrong as @datlife mentioned above. The solution @datlife provided above does not directly solve my problem, a slight change of that did:
> 
> ```
> # add below conv_width_index = K.flatten(K.transpose(conv_width_index))
> # note that at this point conv_dims still follows [height, width]
> enum_width = tf.cast((K.arange(0, stop=conv_dims[1])), dtype=tf.float32)
> enum_width = K.tile(enum_width, [conv_dims[0]])
> enum_width = K.reshape(enum_width, [-1, conv_dims[0], conv_dims[1], 1])
> enum_height = K.cast((K.arange(0, stop=conv_dims[0])), dtype=tf.float32)
> enum_height = K.reshape(enum_height, [-1, 1])
> enum_height = K.tile(enum_height, [1, conv_dims[1]])
> enum_height = K.reshape(enum_height, [-1])
> enum_height = K.reshape(enum_height, [-1, conv_dims[0], conv_dims[1], 1])
> conv_index = tf.stack([enum_width, enum_height], -1)
> 
> # conv_index = K.transpose(K.stack([conv_height_index, conv_width_index]))
> conv_index = K.reshape(conv_index, [1, conv_dims[0], conv_dims[1], 1, 2])
> conv_index = K.cast(conv_index, K.dtype(feats))
> ```
> The basic idea is that you want `conv_index[0, 15, 8, 0, :] == [8, 15]` and `conv_index[0, 12, 6, 0, :] == [6,12]` so that later part `box_xy = (box_xy + conv_index) / conv_dims` will make sense.
> 
> (2) conv_dims need to be fixed to `[width, height]`, solution mentioned in #86 to add one line of code.
> 
> Thanks all for discussing the issue here

Hello,

I've tried to implement this fix but I haven't had any success. I still get predictions that make no sense. I have the hunch that the predicted boxes might be drawn incorrectly but I can't find the error. Did you progress with this error?

Thank you for your comment
